### DISCUSSION TITLE
Is it Possible To Transcribe Twilio Voice Calls in Real-Time with Deepgram using NodeJS?

### DISCUSSION THREAD

**Discussion by titus-shoats** [deepgram/community#307] [General help] [answered]
Hello I've been following the tutorial here https://deepgram.com/learn/deepgram-twilio-streaming to Transcribe Twilio Voice Calls in Real Time. This example is using Python for the code example. I'm interested in knowing can anyone provide the same tutorial using NodeJS. I really dont have any experience in python, as I have experience with JS/NodeJS. There has to be a NodeJS example or demo somewhere, I just cant seem to find one.

If anyone could point in the right direction, I would really appreciate it! Thanks :)

**Comment by DamienDeepgram** [comment]
I never got around to finishing this sample but it’s a great unofficial starting point

https://github.com/DamienDeepgram/twilio-media-stream-live-transcription-node

You might need to leverage keep alive to stop the websocket closing

https://deepgram.com/changelog/stream-keepalive#:~:text=Customers%20can%20now%20keep%20Deepgram,a%20short%20window%20of%20time.

**Reply by titus-shoats** [reply]
@DamienDeepgram im getting an error as soon as the connection is initiated.   Below is the error code.


Listening on Port 8080
New Connection Initiated
A new call has connected.
Starting Media Stream MZ727eb9bd9a060f8636b50f32fa3a0a04
C:\Users\xxxx\Documents\twilio-media-stream-live-transcription-node\index.js:63
              for(let i=0; i<bytes_to_fill; i++){
                             ^

ReferenceError: bytes_to_fill is not defined
    at WebSocket.incoming (C:\Users\xxxx\Documents\twilio-media-stream-live-transcription-node\index.js:63:30)
    at WebSocket.emit (node:events:513:28)
    at Receiver.receiverOnMessage (C:\Users\xxxx\Documents\twilio-media-stream-live-transcription-node\node_modules\ws\lib\websocket.js:1068:20)
    at Receiver.emit (node:events:513:28)
    at Receiver.dataMessage (C:\Users\xxxx\Documents\twilio-media-stream-live-transcription-node\node_modules\ws\lib\receiver.js:517:14)
    at Receiver.getData (C:\Users\xxxx\Documents\twilio-media-stream-live-transcription-node\node_modules\ws\lib\receiver.js:435:17)
    at Receiver.startLoop (C:\Users\xxxx\Documents\twilio-media-stream-live-transcription-node\node_modules\ws\lib\receiver.js:143:22)
    at Receiver._write (C:\Users\xxxx\Documents\twilio-media-stream-live-transcription-node\node_modules\ws\lib\receiver.js:78:10)
    at writeOrBuffer (node:internal/streams/writable:392:12)
    at _write (node:internal/streams/writable:333:10)

Node.js v18.15.0

**Comment by nikolawhallon** [comment, marked as answer]
We also have this version of the Twilio code, written in NodeJS: https://github.com/deepgram/public-code-examples/tree/master/twilio/node hopefully they can help! (it is the analog to the Python code from the blog post)

**Reply by titus-shoats** [reply]
@nikolawhallon im getting an error. When I place the phone call to my twilio #phone number, im getting audio that says "Your application has an error".  Below are the steps and commands I use.

** Im using twilio-proxy-mono.js
17175368XXX is my twilio number

1) twilio phone-numbers:update +17175368XXX --voice-url    https://xxxxx5.ngrok-free.app 
2) npm install
3) npm start

And below is the TwimL Bin im using for proxy mono.  ( I removed xml brackets due to github errors)

Response
    Start
        Stream url=" https://xxxxxx..ngrok-free.app" 
     Start
     Dial>my cell #Dial
Response

**Reply by nikolawhallon** [reply]
Hearing that error usually means the websocket request never reaches the server. I have two questions that might help narrow this down:

What Twilio region are you using?

From the Twilio docs (https://www.twilio.com/docs/voice/twiml/stream): "If communication issues are encountered with your WebSocket server, they will be reported in the Twilio Debugger with additional information about the failure." - is there any info there that you could share?

**Reply by nikolawhallon** [reply]
(I'll note that your url uses https but it should be wss for websocket streams - maybe that's all this is.)

**Reply by titus-shoats** [reply]
@nikolawhallon I finally got it to work successfully  using nodejs. Your advice was very helpful. Thanks
